src/librustc_incremental/persist/fs.rs

```diff
@@ -614,21 +615,15 @@ fn string_to_timestamp(s: &str) -> Result<SystemTime, ()> {
 
 fn crate_path(sess: &Session,
               crate_name: &str,
-              crate_disambiguator: &str)
+              crate_disambiguator: &Fingerprint)
               -> PathBuf {
-    use std::hash::{Hasher, Hash};
-    use std::collections::hash_map::DefaultHasher;
 
     let incr_dir = sess.opts.incremental.as_ref().unwrap().clone();
 
-    // The full crate disambiguator is really long. A hash of it should be
-    // sufficient.
-    let mut hasher = DefaultHasher::new();
-    crate_disambiguator.hash(&mut hasher);
-
+    let crate_disambiguator = crate_disambiguator.to_smaller_hash();
     let crate_name = format!("{}-{}",
                              crate_name,
-                             base_n::encode(hasher.finish(), INT_ENCODE_BASE));
+                             base_n::encode(crate_disambiguator, INT_ENCODE_BASE));
```

Comment:
This truncates the hash to 64 bits, should I implement a 128 but encoding in `base_n`?

Reply:
This is pre-existing?

Reply:
Now sure what you mean exactly, but [`to_smaller_hash()`](https://github.com/rust-lang/rust/blob/c0956ff265d0459e6f643f5454a8560e5dc2712c/src/librustc/ich/fingerprint.rs#L28) already exists and truncates to 64 bits.
I meant to ask if `base_n::encode` should be extended to work with `Fingerprint`/`u128`, in addition to `u64`. To be honest, I saw that it accepts `u64`, so thought that truncating 128->64 might be acceptable in this case

Reply:
The code you wrote should have roughly the same effect as before since `DefaultHasher` also produces 64 bits. We don't want these hashes to be too long because they end up in directory names and long paths/names sometimes cause trouble for various reasons (file system support, operating system API restrictions, etc).

Maybe you could add a comment here, similar to the one that is replaced:
```
// The full crate disambiguator is really long. 64 bits of it should be
// sufficient.
```